### PR TITLE
3016 - cat3 error cleanup

### LINKED ISSUE #3016: Spike - Cat2 Validator Improvement
**Description:**
#2963 makes updates to cat2 validators to include information about the record/field (schema, item number, and friendly name) within the error message. This has a little side effect when the same cat2 validator functions are used within cat3 if/then validators - the item numbers are wrong and the error message contains the extra information for each field and becomes overwhelming/hard to read
- example: `if family affiliation :1 validator1 passed then social security number T2 Item -1 (social security number): 000000000 is in ['000000000', '111111111', '222222222', '333333333', '444444444', '555555555', '666666666', '777777777', '888888888', '999999999'].`

Investigate cleaning up the cat3 error messages. Some possible approaches:
- pass a "ignore_error_message=True` to the cat2 validator from the cat3 validator, have the cat2 validator returned a simplified error or no error (to be filled in by the cat3 validator)
- simply ignore the cat2 validator message and have the cat3 validator create a new message

**Support References**
- #2792
- Errors audit guide - https://app.gitbook.com/o/Yd4Wv0Fi89kSKpmakWrn/s/lIHaHcbpsiyAKVhaisxx/error-style-guide

**Open Questions:**
_Please include any questions, possible solutions or decisions that should be explored during work_ 

- What format should the cat3 validators take?

**Deliverable(s):** 
_Create a list of recommendations or proofs of concept to be achieved to complete this issue_

- [ ] An approach (either completed or outlined in a new ticket) for updating the cat3 validator messages to be more human-readable

### PR DESCRIPTION
**Pull request by jtimpe**
## Summary of Changes
Pull request closes #3016 
* cleans up category 2 error messages when presented "inline" in a category 3 error
   * creates a new `format_error_context` which takes a string `format`. `prefix` is the default, `inline` is passed in anywhere a cat3 validator calls a `make_validator`-created validator.
   * creates a new python `@dataclass` called `ValidatorErrorArguments` - this cleans up the 5 params going into every `make_validator`-created validator

## How to Test
```
cd tdrs-backend && docker-compose up
cd tdrs-frontend && docker-compose up --build
```

1. Open http://localhost:3000/ and sign in.
1. Upload a datafile containing cat2 and cat3 errors
1. Ensure cat2 errors follow the `prefix` format, cat3 errors follow the `inline` format
> *Demo GIF(s) and screenshots for testing procedure*

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
+ [x] update the cat3 validator messages to be more human-readable
+ [x] **`lfrohlich`** and/or **`adpennington`**  confirmed that ACs are met.

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [ ] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [ ] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?

### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

### CODE REVIEW

File: tdrs-backend/tdpservice/parsers/validators.py [hunk 1]
```diff
@@ -115,9 +153,24 @@ def if_then_validator_func(value, row_schema):
             value[result_field_name] if type(value) is dict else getattr(value, result_field_name)
         )
 
+        condition_field = row_schema.get_field_by_name(condition_field_name)
+        result_field = row_schema.get_field_by_name(result_field_name)
+
         # TODO: There is some work to be done here to get the actual friendly name and item numbers of the fields
```
Comment (elipe17) [inline tdrs-backend/tdpservice/parsers/validators.py]: This `TODO` isn't relevant anymore right?

**Review by elipe17 — APPROVED** [review]
Looks good!

**Review by raftmsohani — APPROVED** [review]
LGTM

**Comment by robgendron** [comment]
@reitermb is there an update on this review?

File: tdrs-backend/tdpservice/parsers/test/test_parse.py [hunk 2]
```diff
@@ -331,7 +331,10 @@ def test_parse_bad_trailer_file2(bad_trailer_file_2, dfs):
 
     row_2_error = parser_errors.get(row_number=2)
     assert row_2_error.error_type == ParserErrorCategoryChoices.FIELD_VALUE
-    assert row_2_error.error_message == 'T1: 3 is not larger or equal to 1 and smaller or equal to 2.'
+    assert row_2_error.error_message == (
+        'T1 Item 13 (receives subsidized housing): 3 is not '
```
Comment (victoriaatraft) [inline tdrs-backend/tdpservice/parsers/test/test_parse.py]: Per @reitermb: We can include this in the bigger Cat 3 work if it's scope creepy here but could we convert this to use lists or ranges for the acceptable values? 
 
Part of the broader effort to not use greater than or less than language when the range it's referring to is so small. This is also appearing in a few other places, but otherwise it looks good.

**Comment by victoriaatraft** [comment]
@elipe17 @robgendron @reitermb we added in comments :)

File: tdrs-backend/tdpservice/parsers/test/test_parse.py (hunk 2 above)
Comment (jtimpe) [reply tdrs-backend/tdpservice/parsers/test/test_parse.py]: thanks @victoriaatraft! i believe you've already captured this in #2792, just copying here for posterity/visibility
> If there are any straggler errors that list a greater than or equal to and smaller than or equal to to define the bounds, we should list a simple range or a set of values instead

**Review by victoriaatraft — APPROVED** [review]

**Review by reitermb — APPROVED** [review]
Thanks for the reminder of how we split up this work @jtimpe, sounds good to me!

**Comment by ADPennington** [comment]
@jtimpe after your email notification PR merges can you check into the merge conflict here?

**Comment by ADPennington** [comment]
@jtimpe @reitermb can you confirm for me that this is expected error message configuration for cat 3 types of errors for this ticket? for reference:

- d = develop (reflects previous cat 2 cleanup work)
- p = prod (reflects sprint 101 release)
- q = qasp (this ticket 😄 ) 

3016p2

[testfile.txt](https://github.com/user-attachments/files/16256982/small_Tsection1_cat2.txt)

**Comment by jtimpe** [comment]
@ADPennington @reitermb I pushed a change to remove the extra item name in the second part of the if/then message. that will be overdone by whatever comes of #2792 but i think it cleans up the messaging for now. let me know if anything else in those screenshots looked a little off

**Review by ADPennington — APPROVED** [review]
lgtm @jtimpe 🥇

3016p3

Please note:

- @robgendron per [this comment](https://github.com/raft-tech/TANF-app/pull/2963#pullrequestreview-2154437240), we still need #3049 implemented ASAP to address the remaining byproducts of #2963
